### PR TITLE
Instructor: edit session: add option to discard changes #7232

### LINKED ISSUE #7232: InstructorFeedbackEditPage: Add option to discard changes for session details
**Description of feature/enhancement**

We need to add a "Discard changes" option for the feedback session details too. Precisely speaking, when we try to edit a feedback question,

edit-feedback-3

we get these options -

edit-feedback-4

But when we edit the feedback session details, 

edit-feedback-1

we don't get a "Discard changes" option -

edit-feedback-2

(Icons appear because I tinkered with the code)

**Justification**

If the instructor wants to discard changes in the feedback details, he has no option but to reload the page, but by doing so, they will lose changes made to the feedback questions.

### PR DESCRIPTION
**Pull request by HirdayGupta**
Fixes #7232 

**Old PR:** #7567 
**Reason for new PR:** Incorrect rebasing method resulting in unwanted changes in Old PR.

**Outline of solution:**
- Add a Discard Changes button with the glyphicon ban-circle symbol to the feedbacksession forms header
- Button is visible only when FS is in edit mode and hidden otherwise
- On clicking, confirmation bootbox modal is shown and the callback function for the ok button in the modal resets the Feedback Session Edit form and disables edit mode.
- Ran god mode

**Todo: (Dev help requested)**
- Figure out how to test this feature. 

**Ready for review**

### CODE REVIEW

**Comment by leeyimin** [comment]
@HirdayGupta InstructorFeedbackEditPageUiTest seems like a suitable place to add a test for this! You can refer to `testEditSessionAction` on how to make changes to the form and ensure that clicking the button revert the changes correctly.

File: src/main/webapp/dev/js/main/instructorFeedbackEdit.es6 [hunk 1]
```diff
@@ -1111,6 +1113,15 @@ function setTooltipTriggerOnFeedbackPathMenuOptions() {
 $(document).ready(() => {
     prepareInstructorPages();
 
+    $(document).on('click', '#fsDiscardChanges', () => {
+        const okCallback = function () {
+            document.getElementById('form_feedbacksession').reset();
```
Comment (thenaesh) [inline src/main/webapp/dev/js/main/instructorFeedbackEdit.es6]: Use `$('#form_feedbacksession').trigger('reset');` instead.

File: src/main/webapp/dev/js/main/instructorFeedbackEdit.es6 [hunk 2]
```diff
@@ -1111,6 +1113,15 @@ function setTooltipTriggerOnFeedbackPathMenuOptions() {
 $(document).ready(() => {
     prepareInstructorPages();
 
+    $(document).on('click', '#fsDiscardChanges', () => {
+        const okCallback = function () {
```
Comment (thenaesh) [inline src/main/webapp/dev/js/main/instructorFeedbackEdit.es6]: Use an arrow function here.

Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/instructorFeedbackEdit.es6]: Added arrow function in fe90903. However, the reason I didn't use the arrow function in the first place was that all other `okCallback`s in the same file did not use it. So for consistency, I did not either.

File: src/main/webapp/dev/js/main/instructorFeedbackEdit.es6 (hunk 1 above)
Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/instructorFeedbackEdit.es6]: Done in f2e7a0a. I would like to know why this is preferrable to `.reset()`?

**Comment by HirdayGupta** [comment]
@leeyimin @thenaesh ready for review

File: src/main/webapp/dev/js/main/instructorFeedbackEdit.es6 (hunk 1 above)
Comment (thenaesh) [inline src/main/webapp/dev/js/main/instructorFeedbackEdit.es6]: Mostly for consistency, since we're using JQuery everywhere.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 3]
```diff
@@ -214,10 +214,32 @@ private void testEditSessionAction() throws Exception {
...
+        // Make changes to Feedback session
+        editedSession.setStartTime(Const.TIME_REPRESENTS_NOW);
+        editedSession.setEndTime(Const.TIME_REPRESENTS_LATER);
+        editedSession.setInstructions(new Text("Made some more changes"));
+        editedSession.setGracePeriod(15);
+        feedbackEditPage.editFeedbackSession(editedSession.getStartTime(), editedSession.getEndTime(),
+                                        editedSession.getInstructions(), editedSession.getGracePeriod(), false);
+
+        // Click discard changes button, then cancel and verify html
+        By feedbackSessionEditFormSection = By.id("form_feedbacksession");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: Could do a `verifyHtmlPart` here against "/instructorFeedbackEditDiscardChangesCancel.html" to ensure cancelling the discard modal does not change the edit form.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 4]
```diff
@@ -214,10 +214,32 @@ private void testEditSessionAction() throws Exception {
...
         feedbackEditPage.editFeedbackSession(editedSession.getEndTime(), editedSession.getStartTime(),
-                                        editedSession.getInstructions(), editedSession.getGracePeriod());
+                                        editedSession.getInstructions(), editedSession.getGracePeriod(), true);
 
         String expectedString = "The end time for this feedback session cannot be earlier than the start time.";
         feedbackEditPage.verifyStatus(expectedString);
+
+        ______TS("test discard changes to feedback session");
+
+        // Make changes to Feedback session
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: Could do a page reload here and `verifyHtmlPart` against "/instructorFeedbackEditDiscardChangesOk.html".

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 5]
```diff
@@ -214,10 +214,32 @@ private void testEditSessionAction() throws Exception {
...
+        editedSession.setStartTime(Const.TIME_REPRESENTS_NOW);
+        editedSession.setEndTime(Const.TIME_REPRESENTS_LATER);
+        editedSession.setInstructions(new Text("Made some more changes"));
+        editedSession.setGracePeriod(15);
+        feedbackEditPage.editFeedbackSession(editedSession.getStartTime(), editedSession.getEndTime(),
+                                        editedSession.getInstructions(), editedSession.getGracePeriod(), false);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: Can change more things here, eg. Manual publish time, not send closing reminder emails etc.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 4 above)
Comment (HirdayGupta) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: There are some unexpected differences between the expected and actual HTML when I `verifyHtmlPart` against `/instructorFeedbackEditDiscardChangesOk.html`. You can see the diffs here. https://www.diffchecker.com/8OeggAEz

These seem like they are caused by small differences like the behaviour of `enableEditFS()` in JS as opposed to simply reloading the page.

Requesting advice on best path forward. I was thinking, in this case, if it would be okay to accept a third file to simplify this?

Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: Hmm I would think it would not be good since html files are pretty difficult to check manually. Are there possible workarounds such as checking both pages after clicking edit button? Your solution might not be sufficient as well. If you reload the page at the point at which you check against `instructorFeedbackEditDiscardChangesOk.html`, The `Responses visible from` option changes from `Publish manually` to ` Immediately`.

Comment (HirdayGupta) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: My solution to this was just to reload the page and then `clickEditSessionButton()` followed by `clickSaveSessionButton()`. The changes to the HTML seems to be okay. 

>If you reload the page at the point at which you check against instructorFeedbackEditDiscardChangesOk.html, The Responses visible from option changes from Publish manually to Immediately.

I think this was because of some unrelated change in one of the previous TS

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 5 above)
Comment (HirdayGupta) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: While adding this test, a bug in the feature came to light. Following steps:
- Click edit button
- Make some changes to the FS details
- Click change uncommon email settings link
- Click discard changes --> OK
- Only the changes made to the Feedback Session Instructions field are not discarded.

I can confirm that the changes to FS Instructions field ARE discarded if I do not click the `change uncommon email settings` button. I can also confirm that the same error occurs if I click the `change uncommon visibility options` as well. So there seems to be some problem with how the JS handles these two buttons. Am looking into this.

Comment (HirdayGupta) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: UPDATE: If you notice, when you click `[Change]` alongside either of the Uncommon Settings Info Texts, and the FS is NOT in edit mode, it gets changed to edit mode. I am 95% certain that this is the cause of this unexpected behavior. **However, I cannot find the code that is responsible for this behaviour.**

All I found implementing behaviour of these buttons in `instructorFeedbackEdit.es6` were calls to some methods in its `document.ready` function imported from `common/instructorFeedbacks.es6`. The `.click` function for these buttons is implemented in the `bindUncommonSettingsEvents` function, which looks like this:
```JavaScript
function bindUncommonSettingsEvents() {
    $('#editUncommonSettingsSessionResponsesVisibleButton')
        .click(showUncommonPanelsForSessionResponsesVisible);
    $('#editUncommonSettingsSendEmailsButton')
        .click(showUncommonPanelsForSendEmails);
}
```

The two methods called in the above method look like this:
```JavaScript
function showUncommonPanelsForSessionResponsesVisible() {
    const $sessionResponsesVisiblePanel = $('#sessionResponsesVisiblePanel');

    $('#uncommonSettingsSessionResponsesVisible').after($sessionResponsesVisiblePanel);
    $sessionResponsesVisiblePanel.show();
    $('#uncommonSettingsSessionResponsesVisibleInfoText').parent().hide();
}

function showUncommonPanelsForSendEmails() {
    const $sendEmailsForPanel = $('#sendEmailsForPanel');

    $('#uncommonSettingsSendEmails').after($sendEmailsForPanel);
    $sendEmailsForPanel.show();
    $('#uncommonSettingsSendEmailsInfoText').parent().hide();
}
```

Nowhere in the above code blocks can I find how the FS gets put into Edit Mode if the button is clicked. It will be great if you could help with this @leeyimin.

**Comment by HirdayGupta** [comment]
@leeyimin The only test failing the Travis Build right now `AdminEmailPageUiTest` is passing locally for me (on the same branch). Is there a way to restart the Travis build? Any other way to get green?

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 5 above)
Comment (HirdayGupta) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: @VamsiSangam If you have some time, would it be possible for you to take a look at this and help out? I ask you because I have noticed that you have a knack for JavaScript debugging. Thanks in advance! :)

Comment (VamsiSangam) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackEditPageUiTest.java]: Sure! Will have a look at it when free :smile:

**Comment by VamsiSangam** [comment]
> However, I cannot find the code that is responsible for this behaviour which puts FS into edit mode on click of [Change]

The first line in `readyFeedbackEditPage()` causes it -

``` javascript
$(document).on('click', '.enable-edit-fs', () => enableEditFS());
```

According to this any element with `class="enable-edit-fs"` upon clicking will call `enableEditFs()`. If you inspect the HTML, the edit button, and both `[Change]` links have `enable-edit-fs` class.

---

That being said. I looked at your approach, it revolves around the form `reset`. I didn't know such a thing existed :sweat_smile: . While I was impressed how elegant the resetting thing is, it seems to have a small drawback. If I'm not wrong, resetting a form will reset values as per their HTML tag's value. This has two issues -

- Any value added via JS/AJAX will not be recognized by form reset. [Example](https://codepen.io/vamsisangam/pen/xryjEG)
- If you change attributes (like `data-` attributes) of any element, resetting the form will not reset these values. [Example](https://codepen.io/vamsisangam/pen/yXRjzr?editors=1001) (You'll have to inspect `fname` for this)

These flaws may cause you trouble later on. Which is why I recommend you to follow the existing mechanism which is followed by `Discard changes to question` feature. It is simple too, just store the raw HTML before editing and restore it back if instructor wants to discard changes. Please have a look at the existing mechanism, `backupQuestion()`, `discardChanges()` and `restoreOriginal()` in `instructorFeedbackEdit.es6` is a good place to start. By following the same mechanism you have a 2-fold advantage -

- You'll encounter very few issues, just follow the existing workflow.
- Same goes for testing! :tada: You could look at how `Discard changes to question` feature is tested.

**Comment by leeyimin** [comment]
@HirdayGupta Can you merge changes from master or rebase your changes on top of latest master? It gets difficult to test and review your work if your branch is not using objectify unlike current master.

**Comment by leeyimin** [comment]
@HirdayGupta Ignore the earlier comment. I forgot to update my local branch before commenting 😅

**Comment by HirdayGupta** [comment]
Done @leeyimin

**Comment by HirdayGupta** [comment]
@leeyimin 😆 No worries!

**Comment by leeyimin** [comment]
@HirdayGupta It seems like what @VamsiSangam said is right and `reset` may not be the best way to go about doing this. The edit save action seems to rely on ajax and discarding changes after a edit save action causes the form to revert to its pre-edit-save state, not the current state of the session.

**Comment by HirdayGupta** [comment]
@leeyimin Understood. Working on it 👍

**Comment by leeyimin** [comment]
@HirdayGupta Any update on this PR?

**Comment by HirdayGupta** [comment]
@leeyimin So sorry about the delay on this PR. Have been working on my other PRs, and since this one involved a complete change of approach, the time I spend on this PR ends up being mostly on trial and error. I am working on it though and will update this PR soon. Apologies once again.

**Comment by HirdayGupta** [comment]
@leeyimin I am unable to run the test that is failing on Travis, `FeedbackRubricQuestionUiTest`, on my local machine. It is failing before the line on which Travis throws the `AssertionError`. Here is the stack output from the erroneous run:

```
org.gradle.internal.serialize.PlaceholderException: Element is not clickable at point (1028.1666870117188, 17.166671752929688). Other element would receive the click: <span data-original-title="FRubricQnUiT.instructor" class="text-info tool-tip-decorate" data-toggle="tooltip" title="" data-placement="bottom"></span>
Command duration or timeout: 91 milliseconds
Build info: version: '2.53.1', revision: 'a36b8b1cd5757287168e54b817830adce9b0158d', time: '2016-06-30 19:26:09'
System info: host: 'N/A', ip: 'N/A', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.12.5', java.version: '1.7.0_79'
Session ID: 49e2fe8c-ced5-b643-a67d-2dc868b54caf
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Capabilities [{platform=MAC, acceptSslCerts=true, javascriptEnabled=true, cssSelectorsEnabled=true, databaseEnabled=true, browserName=firefox, handlesAlerts=true, nativeEvents=false, webStorageEnabled=true, rotatable=false, locationContextEnabled=true, applicationCacheEnabled=true, takesScreenshot=true, version=46.0}]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(NativeConstructorAccessorImpl.java:-2)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
        at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:327)
        at org.openqa.selenium.remote.RemoteWebElement.click(RemoteWebElement.java:85)
        at teammates.test.pageobjects.InstructorFeedbackEditPage.moveRubricCol(InstructorFeedbackEditPage.java:404)
        at teammates.test.pageobjects.InstructorFeedbackEditPage.moveRubricColLeft(InstructorFeedbackEditPage.java:382)
        at teammates.test.cases.browsertests.FeedbackRubricQuestionUiTest.testEditQuestionAction(FeedbackRubricQuestionUiTest.java:407)
        at teammates.test.cases.browsertests.FeedbackRubricQuestionUiTest.testEditPage(FeedbackRubricQuestionUiTest.java:179)
        at teammates.test.cases.browsertests.FeedbackRubricQuestionUiTest.allTests(FeedbackRubricQuestionUiTest.java:44)
```

I believe it might have something to do with the merge of #7415. If you look at our `TEAMMATES/master` branch's [commit history](https://github.com/TEAMMATES/teammates/commits/master) that test builds on `AppVeyor` have also been failing because of that test. For some reason, only Travis is able to pass it.

**Comment by HirdayGupta** [comment]
@leeyimin Sorry for not clarifying earlier, the PR is still ongoing. I am almost done with implementing the new approach. I just need to finish figuring out which click actions I need to re-link after re-loading the form from the saved HTML.

**Comment by HirdayGupta** [comment]
@leeyimin Realised another fault while working on this. 

Since the action which is called when edits to FS Details are saved, `InstructorFeedbackEditSaveAction`, returns an Ajax result, the HTML of the web page doesn't change when changes are made. 

As a result both the `backup and restore` approach suggested by @VamsiSangam and the `.trigger('reset')` approach I had originally implemented only restore the HTML to the value that was stored when the page was loaded. 

So essentially, these steps result in an error:
1.  Load the edit page with a certain set of details, and then make some changes and save it. 
2. Click edit again and make some changes. 
3. Now discard the changes (using either approach discussed above), the FS details being displayed don't revert to the changed set of details saved in step 1, but to the original details that the page was loaded with.

This is because the save action returns an Ajax result and doesn't actually update the HTML of the page. The reason `discard changes` works as expected for questions is that the page is reloaded each time the question is saved. 

There are two possible solutions:
1. Get the AJAX result to return all of the data that was updated and then, instead of submitting it using the forms action attribute, I write my own AJAX code to intercept the data and update the HTML manually accordingly. Downside is complexity.
2. Change the edit save action to return a `redirectResult` instead of an `AjaxResult` thereby reloading the page and solving the issue. Downside is that the page gets reloaded each time changes are saved.

**Comment by VamsiSangam** [comment]
According to what you said, I think you are taking backup during page load. Which is correct because the session details are editable on page load. So I think fixing "editable fields after page load" is a better approach. 

What I mean to say is, similar to the feedback questions, we will disable the fields during page load and display an `Edit` button. So here's my suggested approach -

- Disable FS detail fields during page load. Hide `Discard` & `Save` buttons. Display `Edit` button.
- When instructor clicks `Edit` button -
  - Take backup of HTML, i.e., save `fsDetailsBackup = $('#form_feedbacksession').html()`.
  - Hide `Edit`. Display `Discard` & `Save` buttons.
- If the instructor clicks `Discard`, we restore the saved HTML, i.e., `$('#form_feedbacksession').html(fsDetailsBackup)`. Note that this step automatically disables all fields.
- If the instructor clicks `Save` -
  - Send the usual AJAX request.
  - If successful, disable FS details fields and display `Edit` button.
  - If unsuccessful, do nothing (or whatever the current implementation does for this case).

So for the above approach, you'll use existing JS methods `disableEditFS()` & `enableEditFS()`. You might have to make a few changes in these methods. So with this approach I think you'll accomplish the task without any major changes and it does not add to complexity because it is exactly what the FS questions do. I'm not saying it is simple, but you'd probably end up altering lesser code with this approach. :wink:

**Comment by HirdayGupta** [comment]
> According to what you said, I think you are taking backup during page load. Which is correct because the session details are editable on page load. So I think fixing "editable fields after page load" is a better approach.

@VamsiSangam I am taking the backup in the `enableEditFS` method only. When the page loads, if it is to be loaded in edit mode (depending on the presence of the `editsessiondetails` get param), the `enableEditFS` method is called on page load (see line 1094 of [`instructorFeedbackEdit.es6`](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6) ). So technically, **I am taking the backup everytime the edit button is clicked.**  It just so happens that in some cases, a click of the edit button is simulated on page load.

Regarding the " fixing `'editable fields after page load'`", see my PR #7615. We recently changed this to load in edit mode itself assuming that editing FS details was the instructor's intention. I don't think changing this will be necessary for this PR. The actual problem is something else altogether, which I tried to explain in https://github.com/TEAMMATES/teammates/pull/7626#issuecomment-319028375 and will further clarify below:

The issue here is this:

Since the FSEditSave action returns an ajax result, the HTML that is loaded on page load stays the exact same no matter how many edits to the FS details are saved. If you see the [diff here](https://www.diffchecker.com/jdJH94ih), you will see that the HTML of the form upon loading the page and clicking save is _the exact same_ as the HTML after making some edits and then clicking save. The problem is because an Ajax result is returned by the action, the HTML is not updated after saving the changes. This is why the steps I mentioned in https://github.com/TEAMMATES/teammates/pull/7626#issuecomment-319028375 lead to an error. And this error affects both my `.trigger('reset')` method as well as the `backup and restore HTML` method. @VamsiSangam It doesn't matter that I back up the HTMl in the `enableEditFS` method, because the very HTML being backed up is out of date.

The behaviour is different for the questions because the page is roloaded every time a question is added or edited, which means the DOM HTML is always up to date with the latest changes.


> There are two possible solutions:
>1. Get the AJAX result to return all of the data that was updated and then, instead of submitting it using the forms action attribute, I write my own AJAX code to intercept the data and update the HTML manually accordingly. Downside is complexity.
>2. Change the edit save action to return a redirectResult instead of an AjaxResult thereby reloading the page and solving the issue. Downside is that the page gets reloaded each time changes are saved.

And lastly, it is for this problem that I suggested the above two possible solutions in https://github.com/TEAMMATES/teammates/pull/7626#issuecomment-319028375. I hope that clears it up now @VamsiSangam. Awaiting your opinion on this @leeyimin.

**Comment by VamsiSangam** [comment]
Ok, I get it now! Sorry for putting you into the trouble of writing such a detailed explanation :sweat_smile: 

So if HTML backup mechanism doesn't work here, let's change the backup mechanism. I have another suggestion for backup mechanism.

Instead of storing HTML, we can store a dictionary of the input fields. A key-value pair will correspond to a single field in the form. The key will be input name, and the value will be the input value. So your backup function will look something like this -

```js
let fsDetails = {};

function backupFSDetails() {
    // select all form elements
    // exclude button as it is also a form element
    const $fields = $('#form_feedbacksession')
                     .find(':input:enabled')
                     .not('button');

    // must have a fresh dictionary
    fsDetails = {};

    // store the fields in a dictionary
    // key is input element's name
    // value is input element's value
    for (let i = 0; i < $fields.length; i += 1) {
        // .get(i) returns a JS object
        const element = $fields.get(i);

        fsDetails[element.name] = element.value;
    }
}
```

and your restore function would look something like this -

```js

function restoreFSDetails() {
    // put the values back
    for (let inputName in fsDetails) {
        $(`#form_feedbacksession input[name="${inputName}"]`).val(fsDetails[inputName]);
    }
}
```

You need to call `backupFSDetails()` during -
- Page load.
- `Edit` button click.

You need to call `restoreFSDetails()` on `Discard` button click. I came up with this backup mechanism while working on #7759 . I have faith this approach will work in your case. Let me know if there are any problems with this approach.

**Comment by HirdayGupta** [comment]
@VamsiSangam Thank you so much for the new backup mechanism idea. This idea works perfectly as intended for text fields, date fields and dropdowns. However, for radio buttons and checkboxes, the idea is slightly flawed as the `value` attribute holds no meaning there. What we will have to check for is `checked` and `selected` properties and then update them accordingly. This makes for highly convoluted JS code in both the replace and 

An alternate solution is to use the AJAX data being returned. This data returned by the AJAX request contains all of the updated values and updating the fields to match the updated data is a question of replacing these fields with the data returned by the AJAX request. Doing this might also result in some convoluted code, but it would, be adhering with the AJAX principle, that is updating the DOM with returned data. I could also generate the entire form HTML to be replaced on the server side using templates and then simply replace the DOM's form HTML with this.

Since both approaches result in some heavy code writing, I thought it would be best to ask everyone's opinion before I choose an approach. 

@leeyimin @VamsiSangam what do you all think?

**Comment by HirdayGupta** [comment]
@leeyimin requesting your opinion on this.

**Comment by leeyimin** [comment]
@HirdayGupta I personally think that doing a reload is only a temporary fix and ultimately we would want the functionality to work with AJAX requests. (IMO, the same should also apply to question edit save.) 

@thenaesh, hope to get your opinion on this since you are more familiar with the JS architecture!

**Comment by thenaesh** [comment]
@HirdayGupta Personally, I would prefer the saving and restoring code to reside entirely in the JS instead of using templates to generate a chunk of HTML on the server side. This is because I feel the server shouldn't have to concern itself with what is strictly a client-side phenomenon.

To reduce complexity, we can try to save the entire element state instead of picking specific parts (e.g. `val`, `checked`). The [$.extend](https://api.jquery.com/jquery.extend/) method looks promising in this regard. I think restoring can be done with creative use of extend (and passing in `{}` wherever appropriate as well). If all this feels hackish, just raise the level of abstraction by extracting the hackish-looking parts into separate functions.

**Comment by HirdayGupta** [comment]
@thenaesh After extensive experimentation with the `$.extend` jquery method, I have come to the conclusion that it is a little bit of overkill for this use case. While the idea works rather well in theory, and does reduce the complexity of the backup mechanism for different input element types, it is almost impossible to restore with the same reduced complexity. I tried several different ways of updating the DOM with the extended (read: merged) objects, but ran into problems with all the methods including: `element.attr(mergedObj)` and `element.val(mergedObj.val)`. I kept getting inexplicable JQuery errors and was rather befuddled. 😅 

I will instead be creating a new helper class for HTMLBackupAndRestore in which I will handle all the different input types. I will be using a variation of @VamsiSangam's method which he suggested [here](https://github.com/TEAMMATES/teammates/pull/7626#issuecomment-319246937). Expect this to be committed by tonight. Apologies for the horribly long delay in my response to this PR. Too many unexpected things kept happening and I sent nearly all my time debugging our JS code to try and explain what was happening (which is quite nightmarish to debug 😅). This new helper es6 class will come in handy for #7784 as well.

**Comment by VamsiSangam** [comment]
Not sure if this helps. I created a backup like function for one of my PRs, it solves the problem of check-boxes and radio buttons like this - 

https://github.com/VamsiSangam/teammates/blob/c9464c404d6b718ac44bcdd83801b5825919470a/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6#L399-L422

For the instructions, you can do -

``` js
fieldsDict["instructions"] = tinyMCE.activeEditor.getContent();
```

and for restore, you can do a,

```js
tinyMCE.activeEditor.setContent(fieldsDict["instructions"]);
```

The key in the dictionary is the `input` tag's `name` attribute. Every input tag may not have an ID but it will have a `name` attribute. The little catch is that checkboxes and radio buttons have multiple input tags of the same name, so I added a little `if-else` logic in the above code for them. For restoring back values I guess you could do a `$('input[name="${key}"]').val(fieldsDict[key]);`. Not sure if this will work, but maybe you could try this.

**Comment by HirdayGupta** [comment]
Thanks @VamsiSangam I think that should help. Take a look at the HtmlBackupHelper I commit later tonight and let me know what you think. Quick question though: I am having severe trouble restoring the value of a `select` input element. Do you have any idea how to achieve this? Lets say that user makes a change to a select dropdown element and then discards changes. How do I restore the original value of the select tag and get that to display? Have tried setting the `val` to the value I want displayed, but it just doesn't update.

**Comment by VamsiSangam** [comment]
Value of a `select` is restored by using `val()`. [Example](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6#L624) - I wrote that line myself. In that example, we restore changes to the question number `select` field.

Can you tell me the complete jQuery line you are executing to restore the value?

**Comment by HirdayGupta** [comment]
@VamsiSangam thanks for the example! I realised my folly :sweat_smile:

Just one last thing: your above suggestion for the tinyMce editor's content isn't working as tinymce.activeEditor is returning null. I even tried the tinymce.get(Id) method, and that too returned null. (For both of these I tried placing the aforementioned in the richTextEditor.es6 file as well, making new methods). Do you have any suggestions for this? @leeyimin @thenaesh can weigh in here too?

**Comment by VamsiSangam** [comment]
Apologies, my above example works only when the editor is active. But `tinymce.get('instructions').getContent()` and `tinymce.get('instructions').setContent()` seem to work for me -

v2o2wkbf5x

**Comment by HirdayGupta** [comment]
@VamsiSangam It works for me in the console as well, but for some reason, it does not work from the JS code. This line:

```JavaScript
allFieldsDict['instructions'] = tinymce.get('instructions').getContent();
```

Gives me the following error:
``` diff
- instructorFeedbackEdit.js:1 Uncaught TypeError: Cannot read property 'getContent' of null
```

Any suggestions?

**Comment by VamsiSangam** [comment]
I wrote `console.log(tinymce.get('instructions').getContent());` inside `enableEditFS()` and it worked for me. No errors. Can you provide more details as to where you wrote that line of code? If you can create another branch from `HirdayGupta:7232-discard-changes-instructor-edit-fs` branch and push it, then I can see the code in your fork.

**Comment by HirdayGupta** [comment]
@VamsiSangam [This](https://github.com/HirdayGupta/teammates/blob/test-html-backup-1/src/main/webapp/dev/js/common/htmlBackupAndRestoreHelper.es6#L17) is the line where I call `tinymce.get(id).getContent()`

You can see where I call the methods I created in the helper class in `instructorFeedbackEdit.es6` by taking a look at [this diff.](https://github.com/TEAMMATES/teammates/commit/de37b95a87ef1ade00b73825eb78f70f21e6eadd)

**Comment by VamsiSangam** [comment]
Maybe your `htmlBackupAndRestoreHelper` is missing [this](https://github.com/TEAMMATES/teammates/blob/de37b95a87ef1ade00b73825eb78f70f21e6eadd/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6#L1) line in the beginning of the file.

You must check the type of `tinymce` not its ID. I think the check should be written like [this](https://github.com/TEAMMATES/teammates/blob/de37b95a87ef1ade00b73825eb78f70f21e6eadd/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6#L271) instead of [this](https://github.com/HirdayGupta/teammates/blob/test-html-backup-1/src/main/webapp/dev/js/common/htmlBackupAndRestoreHelper.es6#L16).

Other than these two reasons I can't see why your code wouldn't work.

IMO, let's leave creating a helper file for backup and restore. Just write code for `backupFsDetails()` and restoreFsDetails()` methods in `instructorFeedbackEdit.es6` itself. Because this backup mechanism we are following is not perfect, it cannot be used in other places. Eg., the helper methods you wrote can't be used for backup and restore of questions because in questions, options get added/deleted, the whole HTML gets messed up. Your restore method assumes that the input elements stay as they are, but what if they were removed? Since it can't be used at many places it doesn't make sense to create a separate file with helper methods for it.

Few more suggestions -

- `$allFields.get(i)` returns a JS object not jQuery object.
- Use `input` tag's `name` attribute as key, instead of ID, as all elements may or may not have an ID, but are sure to have a `name` attribute.

**Comment by HirdayGupta** [comment]
> IMO, let's leave creating a helper file for backup and restore. Just write code for backupFsDetails() and restoreFsDetails()methods ininstructorFeedbackEdit.es6` itself. Because this backup mechanism we are following is not perfect, it cannot be used in other places. Eg., the helper methods you wrote can't be used for backup and restore of questions because in questions, options get added/deleted, the whole HTML gets messed up. Your restore method assumes that the input elements stay as they are, but what if they were removed? Since it can't be used at many places it doesn't make sense to create a separate file with helper methods for it.

@VamsiSangam, I plan to also work on #7784 after this, where I will work on generalising the helper method I have written to the questions. After which, we can safely add a discard changes button to all our forms (instructorCourseEdit, etc). This is the only reason I wanted to create the helper here. So that I will have a strating point when I get to work on #7784.

> Maybe your htmlBackupAndRestoreHelper is missing this line in the beginning of the file.

Adding that line in didn't fix the error. @VamsiSangam can you tell me exactly where in `instructorFeedbackEdit.es6` had you placed the call when it worked for you? Because it isn't working for me in that file either.

> $allFields.get(i) returns a JS object not jQuery object.

Thanks! Will fix.

> Use input tag's name attribute as key, instead of ID, as all elements may or may not have an ID, but are sure to have a name attribute.

Using name didn't work for checkboxes because different checkboxes had the same name and the dictionary would remember `true` even if only one of the checkboxes with the same name would be checked and the others not. That's why ID works better. And whenever this method needs to be used for any other form, we will either make sure that the id attribute is present or change the code so that if id is not present, it defaults to name.

**Comment by damithc** [comment]
@HirdayGupta any chance of finishing this up during December? Go ahead and close if not possible.

**Comment by wkurniawan07** [comment]
Closing due to an inactivity for a very long period. If anyone is interested to continue the work done here, feel free to reopen/request for reopen.
